source/server-discovery-and-monitoring/tests/unified/pool-clear-checkout-error.json

```diff
@@ -0,0 +1,304 @@
+          "object": "testRunner",
+          "arguments": {
+            "client": "setupClient",
+            "failPoint": {
+              "configureFailPoint": "failCommand",
+              "mode": {
+                "times": 1
+              },
+              "data": {
+                "failCommands": [
+                  "saslContinue"
+                ],
+                "appName": "authErrorTest",
+                "errorCode": 18
+              }
+            }
+          }
+        },
+        {
+          "name": "createEntities",
+          "object": "testRunner",
+          "arguments": {
+            "entities": [
+              {
+                "client": {
+                  "id": "client",
+                  "useMultipleMongoses": false,
+                  "observeEvents": [
+                    "connectionCheckOutStartedEvent",
+                    "poolClearedEvent",
+                    "connectionClosedEvent",
+                    "connectionCheckOutFailedEvent"
+                  ],
+                  "uriOptions": {
+                    "retryWrites": false,
+                    "appname": "authErrorTest"
+                  }
+                }
+              },
+              {
+                "database": {
+                  "id": "database",
+                  "client": "client",
+                  "databaseName": "foo"
+                }
+              },
+              {
+                "collection": {
+                  "id": "collection",
+                  "database": "database",
+                  "collectionName": "bar"
+                }
+              }
+            ]
+          }
+        },
+        {
+          "name": "insertMany",
+          "object": "collection",
+          "arguments": {
+            "documents": [
+              {
+                "_id": 3
+              },
+              {
+                "_id": 4
+              }
+            ]
+          },
+          "expectError": {
+            "isError": true
+          }
+        },
+        {
+          "name": "waitForEvent",
+          "object": "testRunner",
+          "arguments": {
+            "client": "client",
+            "event": {
+              "poolClearedEvent": {}
+            },
+            "count": 1
+          }
+        },
+        {
+          "name": "waitForEvent",
+          "object": "testRunner",
+          "arguments": {
+            "client": "client",
+            "event": {
+              "connectionClosedEvent": {}
+            },
+            "count": 1
+          }
+        }
+      ],
+      "expectEvents": [
+        {
+          "client": "client",
+          "eventType": "cmap",
+          "events": [
+            {
+              "connectionCheckOutStartedEvent": {}
+            },
+            {
+              "poolClearedEvent": {}
+            },
+            {
+              "connectionClosedEvent": {}
+            },
+            {
+              "connectionCheckOutFailedEvent": {}
```

Comment:
We only wait for `connectionClosedEvent`, and don't wait for `connectionCheckOutFailedEvent`, yet `connectionCheckOutFailedEvent` is expected to be delivered the last. This means we have a race condition in this test: `connectionCheckOutFailedEvent` is not guaranteed to be observed. Let's wait for `connectionCheckOutFailedEvent`?

The same applies to the other test in this file, with the description `"Pool is cleared before connection is closed (handshake error)"`.

Reply:
Omission of checkout failed is intentional - these tests are not testing the checkout event emission, but instead testing specifically that the pool cleared error preceeds a connection closed event.

Reply:
If that's the case, then why are these tests required to observe `connectionCheckOutFailedEvent`, and why do they expect `connectionCheckOutFailedEvent`? It seems to me that we should either stop observing and expecting `connectionCheckOutFailedEvent`, or should wait for `connectionCheckOutFailedEvent` instead of / in addition to waiting for `connectionClosedEvent`.